source/math/Quaternion.ooc

```diff
@@ -91,7 +114,73 @@ Quaternion: cover {
 		else
 			roll = asin((upYawPitch scalarProduct(upRotated) * upYawPitch z - upRotated z) / rightYawPitch z)
 
-		FloatRotation3D new(pitch, -yaw, roll)
+		FloatRotation3D new (pitch, -yaw, roll)
+	}
+	distance: func (other: Quaternion) -> Float {
+		(this - other) norm
+	}
+	rotationX: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() < this precision)
+				result = 0.0f
+			else
+				result = (2.0f * (this w * this x + this y * this z)) atan2(1.0f - 2.0f * (this x squared() + this y squared()))
+			result
+		}
+	}
+	rotationY: Float {
+		get {
+			result: Float
+			value := this w * this y - this z * this x
+			if ((value abs() - 0.5f) abs() > this precision)
+				result = ((2.0f * (this w * this y - this z * this x)) clamp(-1, 1)) asin()
+			else
+				result = Float sign(value) * (Float pi / 2.0f)
```

Comment:
Then do that instead. Having `if ((value abs() - 0.5f) abs() > this precision)` in 2 places and `if ((value abs() - 0.5f) abs() < this precision)` in 1 place makes the latter look like a mistake, even though it isn't.